### ISSUE TITLE
Design of Exception Handler on RestControllerAdvice is ugly

### ISSUE THREAD

**Issue by viniciusccarvalho** [spring-cloud/spring-cloud-dataflow#1964]
I was looking into how we handle exceptions on the RestControllerAdvice, and just one method baffled me on how it grew from probably one or two exceptions to this:
```
@ExceptionHandler({ NoSuchStreamDefinitionException.class, NoSuchAppRegistrationException.class,
			NoSuchTaskDefinitionException.class, NoSuchTaskExecutionException.class, NoSuchJobExecutionException.class,
			NoSuchJobInstanceException.class, NoSuchJobException.class, NoSuchStepExecutionException.class,
			MetricsMvcEndpoint.NoSuchMetricException.class, NoSuchAppException.class,
			NoSuchAppInstanceException.class, ApplicationDoesNotExistException.class })
	@ResponseStatus(HttpStatus.NOT_FOUND)
	@ResponseBody
	public VndErrors onNotFoundException(Exception e) {
		String logref = logWarnLevelExceptionMessage(e);
		if (logger.isTraceEnabled()) {
			logTraceLevelStrackTrace(e);
		}
		String msg = getExceptionMessage(e);
		return new VndErrors(logref, msg);
	}
```

Holy Cow! Most of the `NoSuch` exceptions are redundant, some are out of our control, they belong to other libraries, but we really could do better by either having one single NoSuchException type with a parameter of what the heck is missing, and not need to add a new exception to the list every time, or create an abstract type where all NoSuch types could descend from.

@markpollack what do you say?

**Comment by markpollack** [comment]
5 of those are from Spring Batch  (and actually are also in javax.atch) and 1 from Boot so those can not be avoided.  To review, we should look at 

* NoSuchStreamDefinitionException
* NoSuchAppRegistrationException
* NoSuchTaskDefinitionException
* NoSuchTaskExecutionException 
* NoSuchAppException
* NoSuchAppInstanceException
* ApplicationDoesNotExistException

I don't mind find grained exceptions generally speaking.  `NoSuchTaskDefinitionException` and `NoSuchStreamDefinitionException` could perhaps just be consolidated into 'NoSuchDefinitionException`.  `NoSuchAppRegistrationException` seems appropriate given the signatures involved (name and application type).  `NoSuchTaskExecutionException` seems to align with batch's `NoSuchJobExecutionException`.  `NoSuchAppException` and `NoSuchAppInstanceException` seem to overlap and `ApplicationDoesNotExistException` seems to duplicate `NoSuchAppRegistrationException`

So overall, I can see us getting rid of just 3 of these.

**Comment by sabbyanandan** [comment]
Majority of it is already addressed in 2.0 M2 commits; I will create a separate story for the previous comment.
